E2e test fixtures

PeterYinusa opened the pull request:
## Explanation

Fixes: https://github.com/MetaMask/metamask-extension/issues/14340

Previously multiple e2e tests were coupled to a single state.json file.

<img width="485" alt="fixtures" src="https://user-images.githubusercontent.com/53189696/193812740-1011f571-3bcd-4be4-a09c-24c7e2a83294.png">

This PR decouples the state files, and provides an easy to use api to create fixture state, enabling more granular control of each individual tests state.

There are two initial state/fixture objects, default fixture and onboarding fixture. 
The onboarding fixture provides the state of a new user pre onboarding
The default fixture provides the state of an existing user post onboarding


<!--
Thanks for the pull request. Take a moment to answer these questions so that reviewers have the information they need to properly understand your changes:

* What is the current state of things and why does it need to change?
* What is the solution your changes offer and how does it work?

Below is a template to give you some ideas. Feel free to use your own words!

Currently, ...

This is a problem because ...

In order to solve this problem, this pull request ...
-->

## More Information

<!--
Are there any issues, Slack conversations, Zendesk issues, user stories, etc. reviewers should consult to understand this pull request better? For instance:

* Fixes #12345
* See: #67890
-->

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Before

<!-- How did the UI you changed look before your changes? Drag your file(s) below this line: -->

### After

<!-- How does it look now? Drag your file(s) below this line: -->

## Manual Testing Steps

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [ ] PR has been added to the appropriate release Milestone

### + If there are functional changes:

- [ ] Manual testing complete & passed
- [x] "Extension QA Board" label has been applied

Commit:
remove state.json files

Commit:
move file

Commit:
Update Readme

Commit:
Create fixture builder

Commit:
load test fixture

Commit:
remove redundant method

Commit:
update snap tests

Commit:
update stats tests

Commit:
update extension tests

Commit:
update extension tests

PeterYinusa commented on test/e2e/benchmark.js:
```diff
@@ -13,22 +13,26 @@ const {
-    metrics = await driver.collectMetrics();
-  });
+  await withFixtures(
+    { fixtures: new FixtureBuilder().build() },
```
Only the fixture has changed, the rest is just lint formatting

PeterYinusa replied:
Only the fixture has changed, the rest is just lint formatting

Commit:
Update fixture data

PeterYinusa commented on test/e2e/helpers.js:
```diff
@@ -198,26 +198,6 @@ const getWindowHandles = async (driver, handlesCount) => {
   return { extension, dapp, popup };
 };
 
-const connectDappWithExtensionPopup = async (driver) => {
```
We use the FixtureBuilder to add the required state for dapp connections

PeterYinusa commented on test/e2e/lavamoat-stats.js:
```diff
@@ -7,41 +7,45 @@ const yargs = require('yargs/yargs');
-      metrics = parsedLogs.map((pl) => JSON.parse(pl));
-    });
+    await withFixtures(
+      { fixtures: new FixtureBuilder().build() },
```
Only the fixture has changed, the rest is just lint formatting

PeterYinusa commented on test/e2e/mv3-perf-stats/init-load-stats.js:
```diff
@@ -20,42 +21,45 @@ const { withFixtures, tinyDelayMs } = require('../helpers');
-      await driver.delay(1000);
-      const logs = await driver.checkBrowserForLavamoatLogs();
+    await withFixtures(
+      { fixtures: new FixtureBuilder().build() },
```
Only the fixture has changed, the rest is just lint formatting

PeterYinusa commented on test/e2e/mv3-stats.js:
```diff
@@ -20,42 +21,45 @@ const { withFixtures, tinyDelayMs } = require('./helpers');
-      await driver.delay(1000);
-      const logs = await driver.checkBrowserForLavamoatLogs();
+    await withFixtures(
+      { fixtures: new FixtureBuilder().build() },
```
Only the fixture has changed, the rest is just lint formatting

PeterYinusa commented on test/e2e/tests/dapp-interactions.spec.js:
```diff
@@ -34,8 +31,7 @@ describe('Dapp interactions', function () {
         await driver.press('#password', driver.Key.ENTER);
 
-        // Connect to Dapp0
-        await connectDappWithExtensionPopup(driver, 0);
```
We do not actually need to connect to the dapp to click the add ethereum chain button

PeterYinusa commented on test/e2e/tests/dapp-interactions.spec.js:
```diff
@@ -78,8 +76,7 @@ describe('Dapp interactions', function () {
         await driver.press('#password', driver.Key.ENTER);
 
-        // Connect to Dapp0
-        await connectDappWithExtensionPopup(driver, 0);
```
We can use the fixture to connect to the dapp

PeterYinusa commented on test/e2e/tests/edit-gas-fee.spec.js:
```diff
@@ -34,7 +39,7 @@ describe('Editing Confirm Transaction', function () {
         );
         const transactionAmount = transactionAmounts[0];
-        assert.equal(await transactionAmount.getText(), '2.2');
+        assert.equal(await transactionAmount.getText(), '1');
```
The value has changed as we are now reusing an existing transaction

PeterYinusa commented on test/e2e/tests/edit-gas-fee.spec.js:
```diff
@@ -112,7 +122,7 @@ describe('Editing Confirm Transaction', function () {
         );
         const transactionAmount = transactionAmounts[0];
-        assert.equal(await transactionAmount.getText(), '2.2');
+        assert.equal(await transactionAmount.getText(), '1');
```
The value has changed as we are now reusing an existing transaction

PeterYinusa commented on test/e2e/tests/edit-gas-fee.spec.js:
```diff
@@ -194,7 +209,7 @@ describe('Editing Confirm Transaction', function () {
         await driver.press('#password', driver.Key.ENTER);
 
         // open dapp and connect
-        await connectDappWithExtensionPopup(driver);
```
We can use the fixture to connect to the dapp

PeterYinusa commented on test/e2e/tests/from-import-ui.spec.js:
```diff
@@ -315,7 +344,6 @@ describe('MetaMask Import UI', function () {
         const importJsonFile = path.join(
           __dirname,
           '..',
-          'fixtures',
```
moved this file out of the fixture folder as its not a fixture

Commit:
snap test dapp connection

PeterYinusa commented on test/e2e/snaps/test-snap-bip-44.spec.js:
```diff
@@ -39,26 +42,9 @@ describe('Test Snap bip-44', function () {
         // connect the snap
         await driver.clickElement('#connectBip44');
 
-        // switch to metamask extension and click connect
```
Connect to the snap test dapp in the fixture

PeterYinusa commented on test/e2e/snaps/test-snap-confirm.spec.js:
```diff
@@ -31,26 +34,9 @@ describe('Test Snap Confirm', function () {
         await driver.clickElement('#connectHello');
 
-        // switch to metamask extension and click connect
-        await driver.waitUntilXWindowHandles(2, 5000, 10000);
```
Connect to the snap test dapp in the fixture

PeterYinusa commented on test/e2e/snaps/test-snap-error.spec.js:
```diff
@@ -32,25 +35,8 @@ describe('Test Snap Error', function () {
         await driver.fill('#snapId2', 'npm:@metamask/test-snap-error');
         await driver.clickElement('#connectError');
 
-        // switch to metamask extension and click connect
```
Connect to the snap test dapp in the fixture

PeterYinusa commented on test/e2e/snaps/test-snap-managestate.spec.js:
```diff
@@ -40,25 +43,9 @@ describe('Test Snap manageState', function () {
         // connect the snap
         await driver.clickElement('#connectManageState');
 
-        // switch to metamask extension and click connect
```
Connect to the snap test dapp in the fixture

seaona commented on test/e2e/fixture-builder.js:
```diff
@@ -0,0 +1,647 @@
+      AddressBookController: {},
+      AlertController: {},
+      AnnouncementController: {
+        announcements: {
```
I have a couple of suggestions for the announcements part. Since it's pretty hard to tell what each number represents, I would maybe add a comment on each one, with a name indicating which announcement represent. (In the code it's also hard to find, as it's displayed in the same way as here).

I've seen that there are 14 announcements. Do you think we should include all of them here? We are missing `0, 2, 12, 13 `

![image](https://user-images.githubusercontent.com/54408225/193815562-1e0d6f56-1be0-4ccc-bb42-ea04df197ab9.png)

PeterYinusa replied:
Yes, it should probably go all the way up to 15 for the latest notification about the merge

Commit:
Update fixture data

Gudahtt commented on test/e2e/fixture-builder.js:
```diff
@@ -0,0 +1,659 @@
+class Fixture {
```
Nit: Not sure that making this a `class` is helpful. We aren't really using any class features, aside from that the default fixture state is "protected" against accidental modification.

We can more simply protect it by making this a function that returns this object. Or by making it a constant and running `deep-freeze-strict` on it.

Gudahtt commented on test/e2e/tests/from-import-ui.spec.js:
```diff
@@ -27,7 +28,16 @@ describe('MetaMask Import UI', function () {
     await withFixtures(
       {
-        fixtures: 'onboarding',
+        fixtures: new FixtureBuilder()
```
Hmm. So you've recreated something that is functionally equivalent to a pre-onboarding state for this test. But there are still _a lot_ of differences.

Instead of initializing with post-onboarding state and then undoing parts of it, we could have multiple default states (pre and post onboarding). Or we could move onboarding to `withOnboarding` and call it explicitly in most places.

What you have here still works. But it would be better if the state was closer to what it would be in practice.

Gudahtt commented on test/e2e/tests/from-import-ui.spec.js:
```diff
@@ -192,7 +211,12 @@ describe('MetaMask Import UI', function () {
+        fixtures: new FixtureBuilder()
+          .withKeyringController({
+            vault:
+              '{"data":"Ot+BTtJPag0xubdiv1nO9bsSvTHivHCd6CD7Lxgb1McYw3VqMjgp5rPMZmblJ1lscuMxyiqp99G52uXO9S0em6F9htpa+t/wn6qubRKTTNG9fxNzQrKXRDNhdgfYckVk5VAZ4fgl2iMZcRDvS8H/+gucVKJ33Sl6mXyPofdexXhWDCU6uR2YecnfaIum9cL2u/GqOMPE3jxzy0Wip0x2Jyp3QOKhvu8A3GIjzagLOaQ7a1APdl8=","iv":"lbsyPeGYWU6U1+jvmW9UHg==","salt":"Zmbhpskwxe4rYfXtELBvlcvW4HISPBATRmMqzsnZPMg="}',
```
Nit: It would be nice to move this to a variable to avoid repeating it, and to add some explanation for what this is (e.g. `vaultWithSingleAccount` or something like that, whatever this is).

PeterYinusa replied:
Will work on all these suggestions

Commit:
add onboarding fixture

Commit:
use onboarding fixture

Commit:
reuse import account vault

Commit:
remove unnecessary use of class

Commit:
use fixture builder in new tests

Commit:
switch to function

Commit:
update default fixture

Commit:
update default fixture

Commit:
update test

Commit:
update 1559 test fixttures

Commit:
update 1559 test fixtures

Commit:
update 1559 test fixtures

Commit:
dismiss 3box whats new

PeterYinusa commented on test/e2e/snaps/test-snap-update.spec.js:
```diff
@@ -36,30 +39,13 @@ describe('Test Snap update', function () {
           'MetaMask Notification',
           windowHandles,
         );
-        await driver.clickElement(
```
Connect to the snap test dapp in the fixture

PeterYinusa commented on test/e2e/tests/failing-contract.spec.js:
```diff
@@ -18,7 +19,19 @@ describe('Failing contract interaction ', function () {
         dapp: true,
-        fixtures: 'connected-state',
+        fixtures: new FixtureBuilder()
+          .withNetworkController({
```
The send failing transaction is a type 2 transaction

PeterYinusa commented on test/e2e/tests/gas-api-fallback.spec.js:
```diff
@@ -29,7 +30,15 @@ describe('Gas API fallback', function () {
       {
-        fixtures: 'imported-account',
+        fixtures: new FixtureBuilder()
+          .withNetworkController({
```
This transaction is on a 1559 enabled network

PeterYinusa commented on test/e2e/fixture-builder.js:
```diff
@@ -0,0 +1,1002 @@
+        ? this.fixture.data.AddressBookController
+        : (this.fixture.data.AddressBookController = {}),
+      data,
+    );
```
The following controllers were removed from either/both the onboarding fixture and the default fixture, so we set it as an empty object if it does not exist:
AddressBookController, CollectiblesController, IncomingTransactionsController, PermissionLogController, ThreeBoxController, TokenListController, TransactionController

Gudahtt commented:
Great work @PeterYinusa ! I'm still partway through reviewing this, but I thought I'd share something that has been helping me understand the changes to state fixtures.

I added a feature recently to the `prettier-plugin-sort-json` package to support recursive sorting, which makes the state fixtures easier to compare. To run it locally, update `prettier-plugin-sort-json` to v0.0.3 then run this command: `yarn prettier --write --json-recursive-sort=true ./test/e2e/fixtures/**/*.json`. Then they will all be easier to compare! I am using the command `code --diff ./test/e2e/fixtures/import-ui/state.json ./test/e2e/fixtures/<fixture to compare>/state.json` to compare them.

Gudahtt commented on test/e2e/tests/failing-contract.spec.js:
```diff
@@ -18,7 +19,19 @@ describe('Failing contract interaction ', function () {
+            },
+          })
+          .withPermissionControllerConnectedToTestDapp()
+          .withPreferencesController({
+            eip1559V2Enabled: true,
+          })
```
This this part needed? It doesn't seem to be in the `connected-state` fixture

```suggestion
```

Gudahtt commented on test/e2e/fixtures/README.md:
```diff
@@ -1,10 +1,3 @@
-
-1. A `state.json` file that represents a the saved state for the extension (see _Generating fixture data_ below)
-
 ## Generating fixture data
```
Maybe this should be made the main title of this file, and moved into `docs/` as an independent article on generating fixture data. It's not very discoverable left here in this directory

Gudahtt commented on test/e2e/fixture-server.js:
```diff
@@ -94,15 +92,11 @@ class FixtureServer {
-      const data = await fs.readFile(statePath);
-      const rawState = JSON.parse(data.toString('utf-8'));
       state = performStateSubstitutions(rawState);
       this._initialStateCache.set(statePath, state);
```
The caching here doesn't really make sense anymore. It's probably only resulting in increased memory usage, because we're still getting a new state object for each test. This state cache is now nothing more than a memory leak.

I think we can remove this caching logic, along with the `statePath` parameter.

PeterYinusa replied:
Agreed

Gudahtt commented on test/e2e/tests/add-hide-token.spec.js:
```diff
@@ -69,7 +86,21 @@ describe('Add existing token using search', function () {
     await withFixtures(
       {
-        fixtures: 'imported-account',
+        fixtures: new FixtureBuilder()
+          .withTokenListController({
+            tokenList: {
+              '0x0D8775F648430679A709E98d2b0Cb6250d2887EF': {
+                address: '0x0D8775F648430679A709E98d2b0Cb6250d2887EF',
+                decimals: 18,
+                iconUrl:
+                  'https://static.metaswap.codefi.network/api/v1/tokenIcons/1337/0x0d8775f648430679a709e98d2b0cb6250d2887ef.png',
+                name: 'Basic Attention Token',
+                occurrences: null,
+                symbol: 'BAT',
+              },
+            },
+          })
+          .build(),
```
Is this needed? The old fixture didn't have it. I'm guessing it's using the fallback `contract-metadata` token list.

```suggestion
        fixtures: new FixtureBuilder().build(),
```

PeterYinusa replied:
Good catch, this was unnecessary as we are now mocking the token list, I've enabled token detection bringing this test in line with how it use to work

Gudahtt reviewed (commented):
I've finished my first pass. The only big concern I had was the caching in the fixture server, I think that has to go. I left a few other suggestions but nothing too critical. Everything else looks great!

Commit:
remove redundant code

Commit:
move docs

PeterYinusa commented:
> I've finished my first pass. The only big concern I had was the caching in the fixture server, I think that has to go. I left a few other suggestions but nothing too critical. Everything else looks great!

Thank you, appreciate your review and feedback

Commit:
remove unused code

Commit:
token detection

Commit:
use default timeout

Commit:
remove redundant code

PeterYinusa commented:
> Great work @PeterYinusa ! I'm still partway through reviewing this, but I thought I'd share something that has been helping me understand the changes to state fixtures.
> 
> I added a feature recently to the `prettier-plugin-sort-json` package to support recursive sorting, which makes the state fixtures easier to compare. To run it locally, update `prettier-plugin-sort-json` to v0.0.3 then run this command: `yarn prettier --write --json-recursive-sort=true ./test/e2e/fixtures/**/*.json`. Then they will all be easier to compare! I am using the command `code --diff ./test/e2e/fixtures/import-ui/state.json ./test/e2e/fixtures/<fixture to compare>/state.json` to compare them.

Great idea, I'll give the fixtures a once over before asking for it to be reviewed again. I know t least one test fixture may be wonky at the moment e.g. the identities in the from-import-ui.spec.js

Gudahtt reviewed (dismissed):
LGTM!

Commit:
Update fixture builder

hide `Protect your funds` dialog
remove browser environment
remove default network details
hide dismiss seed backup reminder
recursively merges fixture data

Commit:
add token to tokencontroller

Commit:
remove network details

Commit:
add missing identities to preference controller

Commit:
remove duplicate properties

Commit:
update bip-32 to use fixturebuilder

Commit:
alphabetise snap permissions

Commit:
update get snaps to use fixturebuilder

Commit:
Update test-snap-bip-32.spec.js

wait for window

PeterYinusa commented on test/e2e/snaps/test-snap-bip-32.spec.js:
```diff
@@ -36,26 +39,9 @@ describe('Test Snap bip-32', function () {
         await driver.fill('#snapId6', 'npm:@metamask/test-snap-bip32');
         await driver.clickElement('#connectBip32');
 
-        // switch to metamask extension and click connect
```
Connect to the snap test dapp in the fixture

seaona commented on test/e2e/tests/add-hide-token.spec.js:
```diff
@@ -69,12 +95,11 @@ describe('Add existing token using search', function () {
       {
-        fixtures: 'imported-account',
+        fixtures: new FixtureBuilder()
+          .withPreferencesController({ useTokenDetection: true })
```
was wondering whether we need `useTokenDetection` here?

PeterYinusa replied:
Good question.
The current imported account fixture on develop has token detection switched on. See [here](https://github.com/MetaMask/metamask-extension/blob/65e3047a748d2a9b428959801d098831d1d7c532/test/e2e/fixtures/imported-account/state.json#L149) 
Initially I had token detection off, so had to add the token to state. See [here](https://github.com/MetaMask/metamask-extension/pull/16061#discussion_r993968925)
But after discussion with Mark above, I realised it's better to remove this extra state and enable token detection to bring the test inline with how it was before.
So with token detection switched on, the extension will now call the token api which is mocked. See [here](https://github.com/MetaMask/metamask-extension/blob/65e3047a748d2a9b428959801d098831d1d7c532/test/e2e/mock-e2e.js#L153)

jpuri reviewed (dismissed):
LGTM

Gudahtt reviewed (dismissed):
LGTM!

Commit:
add popular network state

Commit:
update test

Commit:
lint

PeterYinusa commented on test/e2e/tests/chain-interactions.spec.js:
```diff
@@ -99,7 +100,9 @@ describe('Chain Interactions', function () {
-        const networkDisplay = await driver.findElement('.network-display');
+        const networkDisplay = await driver.findElement(
+          '[data-testid="network-display"]',
+        );
```
Nothings changed here, this is the same element as the original, just using a different selector

Gudahtt reviewed (approved):
LGTM!
